contracts/RandomBeaconHistory.cdc

```diff
@@ -36,25 +49,170 @@ access(all) contract RandomBeaconHistory {
+            }
+            let backfiller = RandomBeaconHistory.borrowBackfiller() ?? panic("Problem borrowing backfiller")
+
+            // check for any existing gap and backfill using the input random source if needed.
+            backfiller.backfill(randomSource: randomSourceHistory)
 
+            // we are now at the correct index to record the source of randomness
+            // created by the protocol for the current block
             RandomBeaconHistory.randomSourceHistory.append(randomSourceHistory)
         }
     }
 
+    /* --- Backfiller --- */
+    //
+    /// A recovery mechanism designed to backfill missed sources of randomness in the event of a missed commitment due
+    /// to a system transaction failure.
+    ///
+    access(all) resource Backfiller {
+        /// Start index of the first gap in the `randomSourceHistory` array where random sources were not recorded,
+        /// because of a heartbeat failure.
+        /// There may be non contiguous gaps in the history, `gapStartIndex` is the start index of the lowest-height
+        /// gap.
+        /// If no gaps exist, `gapStartIndex` is equal to the `randomSourceHistory` array length.
+        access(contract) var gapStartIndex: UInt64
+        /// BackFilling is limited to a maximum number of entries per call to limit the computation cost.
+        /// This means a large gap may need a few calls to get fully backfilled.
+        access(contract) var maxEntriesPerCall: UInt64
+
+        init() {
+            self.gapStartIndex = UInt64(RandomBeaconHistory.randomSourceHistory.length)
+            self.maxEntriesPerCall = 100
+        }
+
+        access(all) view fun getMaxEntriesPerCall() : UInt64 {
+            return self.maxEntriesPerCall
+        }
+
+        access(all) fun setMaxEntriesPerCall(max: UInt64) {
+            assert (
+                max > 0,
+                message : "the maximum entry per call must be strictly positive"
+            )
+            self.maxEntriesPerCall = max
+        }
+
+        /// Finds the correct index to fill with the new random source. If a gap is detected, emits the
+        /// RandomHistoryMissing event.
+        ///
+        access(contract) view fun findGapAndReturnCorrectIndex(): UInt64 {
+
+            let currentBlockHeight = getCurrentBlock().height
+            // correct index to fill with the new random source
+            // so that eventually randomSourceHistory[correctIndex] = inputRandom
+            let correctIndex = currentBlockHeight - RandomBeaconHistory.lowestHeight!
+
+            // cache the array length in a local variable to avoid multiple calls to `length`
+            // when the array is too large
+            var arrayLength = UInt64(RandomBeaconHistory.randomSourceHistory.length)
+
+            // if a new gap is detected, emit an event
+            if correctIndex > arrayLength {
```

Comment:
Since accessing the length is constant time and cheap, I added this minor optimization FYI https://github.com/onflow/flow-core-contracts/pull/416/commits/e913f3aa42484ff822b844027b4a531f35fba4e7